tgui/packages/tgui/interfaces/ColorPickerModal.tsx

```diff
@@ -0,0 +1,766 @@
+      onChange({
+        s: interaction.left * 100,
+        v: 100 - interaction.top * 100,
+      });
+    };
+
+    const handleKey = (offset: Interaction) => {
+      onChange({
+        s: clamp(hsva.s + offset.left * 100, 0, 100),
+        v: clamp(hsva.v - offset.top * 100, 0, 100),
+      });
+    };
+
+    const containerStyle = useMemo(
+      () => ({
+        backgroundColor: hsvaToHslString({
+          h: hsva.h,
+          s: 100,
+          v: 100,
+          a: 1,
+        }),
+      }),
+      [hsva.h],
+    );
+
+    return (
+      <div className="react-colorful__saturation_value" style={containerStyle}>
+        <Interactive
+          onMove={handleMove}
+          onKey={handleKey}
+          aria-label="Color"
+          aria-valuetext={`Saturation ${Math.round(
+            hsva.s,
+          )}%, Brightness ${Math.round(hsva.v)}%`}
+        >
+          <Pointer
+            className="react-colorful__saturation_value-pointer"
+            top={1 - hsva.v / 100}
+            left={hsva.s / 100}
+            color={hsvaToHslString(hsva)}
+          />
+        </Interactive>
+      </div>
+    );
+  },
+);
+
+interface HueProps {
+  className?: string;
+  hue: number;
+  onChange: (newHue: Partial<HsvaColor>) => void;
+}
+
+const Hue: React.FC<HueProps> = React.memo(({ className, hue, onChange }) => {
+  const handleMove = (interaction: Interaction) => {
+    onChange({ h: 360 * interaction.left });
+  };
+
+  const handleKey = (offset: Interaction) => {
+    onChange({ h: clamp(hue + offset.left * 360, 0, 360) });
+  };
+
+  const nodeClassName = classes(['react-colorful__hue', className]);
+
+  return (
+    <div className={nodeClassName}>
+      <Interactive
+        onMove={handleMove}
+        onKey={handleKey}
+        aria-label="Hue"
+        aria-valuenow={Math.round(hue)}
+        aria-valuemax={360}
+        aria-valuemin={0}
+      >
+        <Pointer
+          className="react-colorful__hue-pointer"
+          left={hue / 360}
+          color={hsvaToHslString({ h: hue, s: 100, v: 100, a: 1 })}
+        />
+      </Interactive>
+    </div>
+  );
+});
+
+interface SaturationProps {
+  className?: string;
+  color: HsvaColor;
+  onChange: (newSaturation: Partial<HsvaColor>) => void;
+}
+
+const Saturation: React.FC<SaturationProps> = React.memo(
+  ({ className, color, onChange }) => {
+    const handleMove = (interaction: Interaction) => {
+      onChange({ s: 100 * interaction.left });
+    };
+
+    const handleKey = (offset: Interaction) => {
+      onChange({ s: clamp(color.s + offset.left * 100, 0, 100) });
+    };
+
+    const nodeClassName = classes(['react-colorful__saturation', className]);
+
+    const background = useMemo(
+      () =>
+        `linear-gradient(to right, ${hsvaToHslString({
+          h: color.h,
+          s: 0,
+          v: color.v,
+          a: 1,
+        })}, ${hsvaToHslString({ h: color.h, s: 100, v: color.v, a: 1 })})`,
```

Comment:
Why is background a string? Didn't react style prop support objects?

Reply:
The background is indeed a string because we’re using the linear-gradient CSS function. React’s style prop supports objects for simple properties like backgroundColor or color. However, for complex CSS functions like linear-gradient(...), it must be provided as a string. This is necessary because linear-gradient cannot be expressed as a JS object – it’s a CSS-specific syntax.